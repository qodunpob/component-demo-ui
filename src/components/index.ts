export * from './gitHubLink'
export * from './preview'
export * from './propList'
export * from './section'
