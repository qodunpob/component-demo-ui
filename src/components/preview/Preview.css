.component-demo-ui-preview {
  display: flex;
  flex-direction: column;
  gap: 1.25em;
}

.component-demo-ui-title {
  margin-bottom: 0;
  font-size: 32px;
  font-weight: 300;
}

.component-demo-ui-caption {
  margin: 0;
}
