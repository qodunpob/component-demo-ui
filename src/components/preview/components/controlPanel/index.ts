export * from './ControlPanel'
export * from './ControlPanel.types'
