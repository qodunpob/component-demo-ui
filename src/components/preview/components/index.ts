export * from './controlPanel'
export * from './inspector'
