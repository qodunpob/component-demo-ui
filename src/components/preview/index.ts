export * from './Preview'
export * from './Preview.types'
export type { ControlDefinition } from './components'
