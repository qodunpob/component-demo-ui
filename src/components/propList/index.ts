export * from './PropList'
export * from './PropList.types'
