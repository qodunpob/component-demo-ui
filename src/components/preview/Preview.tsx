import React, { forwardRef, type ReactElement } from 'react'
import './Preview.css'
import { type ControlDefinition, ControlPanel, Inspector } from './components'
import { type PreviewRef } from './Preview.types'
import { useControlPanel, useInspector } from './hooks'
import { isNotEmptyArray, isNotEmptyString } from '../../utils/common.utils'
import cn from 'classnames'

export interface PreviewProps {
  className?: string
  title: string
  children: ReactElement
  caption?: string
  description?: string
  controls?: ControlDefinition[]
  showInspector?: boolean
  inspectorTitle?: string
}

export const Preview = forwardRef<PreviewRef, PreviewProps>(
  (
    {
      className,
      title,
      caption,
      description,
      controls,
      showInspector = false,
      inspectorTitle,
      children,
      ...restProps
    },
    ref
  ) => {
    const { component, handleControlsChange } = useControlPanel({
      children,
      controls
    })
    const { data } = useInspector(ref)
    return (
      <div
        className={cn(
          className,
          'component-demo-ui-container',
          'component-demo-ui-preview'
        )}
        {...restProps}
      >
        <h1 className='component-demo-ui-preview-title'>{title}</h1>
        {isNotEmptyString(caption) && (
          <p className='component-demo-ui-preview-caption'>{caption}</p>
        )}
        {component}
        {isNotEmptyArray(controls) && (
          <ControlPanel controls={controls} onChange={handleControlsChange} />
        )}
        {showInspector && <Inspector title={inspectorTitle} data={data} />}
        {isNotEmptyString(description) && (
          <p className='component-demo-ui-preview-description'>{description}</p>
        )}
      </div>
    )
  }
)

Preview.displayName = 'Preview'
