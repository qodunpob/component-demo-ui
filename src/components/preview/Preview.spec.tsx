import React, { useEffect, useRef } from 'react'
import { Preview, type PreviewProps } from './Preview'
import {
  fireEvent,
  render,
  type RenderResult,
  screen
} from '@testing-library/react'
import { type ControlDefinition } from '~/models'
import { type PreviewRef } from '~/components'

describe('Preview', () => {
  it('should display a preview of the component', () => {
    renderComponent()
    const helloWorld = screen.queryByTestId('hello-world')
    expect(helloWorld).toMatchSnapshot()
  })

  it('should display caption and description', () => {
    const { baseElement } = renderComponent({
      caption: 'Caption',
      description: 'Description'
    })
    expect(baseElement).toMatchSnapshot()
  })

  it('should display controls', () => {
    renderComponent({ controls })
    const controlPanel = screen.getByTestId('component-demo-ui-controls')
    expect(controlPanel).toMatchSnapshot()
  })

  it('should change the state of the component using controls', () => {
    renderComponent({ controls })

    const helloWorld = screen.queryByTestId('hello-world')
    expect(helloWorld).toMatchSnapshot()

    const greetingControl = screen.getByTestId<HTMLSelectElement>(
      'component-demo-ui-control-greeting'
    )
    fireEvent.change(greetingControl, { target: { value: '„Åì„Çì„Å´„Å°„ÅØ' } })

    const nameControl = screen.getByTestId<HTMLInputElement>(
      'component-demo-ui-control-name'
    )
    fireEvent.change(nameControl, { target: { value: 'David' } })

    const hasEmojiControl = screen.getByTestId<HTMLInputElement>(
      'component-demo-ui-control-hasEmoji'
    )
    fireEvent.click(hasEmojiControl)

    expect(helloWorld).toMatchSnapshot()
  })

  it('should display the inspector', () => {
    renderComponent({ showInspector: true, inspectorTitle: 'Inspector' })
    const inspector = screen.getByTestId('component-demo-ui-inspector')
    expect(inspector).toMatchSnapshot()
  })

  it('should provide an interface to interact with the inspector', () => {
    renderComponent({
      showInspector: true,
      inspectorTitle: 'Inspector',
      onReady: (ref) => {
        ref.inspect({ hello: 'World' })
      }
    })
    const inspector = screen.getByTestId('component-demo-ui-inspector')
    expect(inspector).toMatchSnapshot()
  })
})

interface HelloWorldProps {
  greeting?: 'Hello' | 'Bonjour' | '„Åì„Çì„Å´„Å°„ÅØ'
  name?: string
  hasEmoji?: boolean
}

const HelloWorld: React.FC<HelloWorldProps> = ({
  greeting = 'Hello',
  name = 'World',
  hasEmoji = false
}) => (
  <div data-testid='hello-world'>
    {`${greeting}, ${name}`} {hasEmoji && 'üëã'}
  </div>
)

interface WrapperProps extends Partial<Omit<PreviewProps, 'children'>> {
  onReady?: (ref: PreviewRef) => void
}

const Wrapper: React.FC<WrapperProps> = ({
  title = 'Preview',
  onReady,
  ...restProps
}) => {
  const ref = useRef<PreviewRef>(null)
  useEffect(() => {
    if (ref.current != null) {
      onReady?.(ref.current)
    }
  }, [onReady])

  return (
    <Preview ref={ref} title={title} {...restProps}>
      <HelloWorld />
    </Preview>
  )
}

const renderComponent = (props: WrapperProps = {}): RenderResult =>
  render(<Wrapper {...props} />)

const controls: ControlDefinition[] = [
  {
    name: 'greeting',
    type: 'select',
    options: ['Hello', 'Bonjour', '„Åì„Çì„Å´„Å°„ÅØ'],
    initialValue: 'Bonjour'
  },
  {
    name: 'name',
    type: 'text',
    initialValue: 'Bob'
  },
  {
    name: 'hasEmoji',
    type: 'checkbox',
    initialValue: true
  }
]
