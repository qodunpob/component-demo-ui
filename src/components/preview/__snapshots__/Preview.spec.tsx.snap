// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Preview should change the state of the component using controls 1`] = `
<div
  data-testid="hello-world"
>
  Bonjour, Bob
   
  üëã
</div>
`;

exports[`Preview should change the state of the component using controls 2`] = `
<div
  data-testid="hello-world"
>
  „Åì„Çì„Å´„Å°„ÅØ, David
   
</div>
`;

exports[`Preview should display a preview of the component 1`] = `
<div
  data-testid="hello-world"
>
  Hello, World
   
</div>
`;

exports[`Preview should display caption and description 1`] = `
<body>
  <div>
    <div
      class="component-demo-ui-container component-demo-ui-preview"
      data-testid="component-demo-ui-preview"
    >
      <h1
        class="component-demo-ui-preview-title"
      >
        Preview
      </h1>
      <p
        class="component-demo-ui-preview-caption"
      >
        Caption
      </p>
      <div
        data-testid="hello-world"
      >
        Hello, World
         
      </div>
      <p
        class="component-demo-ui-preview-description"
      >
        Description
      </p>
    </div>
  </div>
</body>
`;

exports[`Preview should display controls 1`] = `
<div
  class="component-demo-ui-control-panel"
  data-testid="component-demo-ui-controls"
>
  <h3
    class="component-demo-ui-control-panel-title"
  >
    Controls
  </h3>
  <div
    class="component-demo-ui-control-panel-label"
  >
    greeting
  </div>
  <div
    class="component-demo-ui-control-panel-control"
  >
    <select
      class="component-demo-ui-control component-demo-ui-select-control"
      data-testid="component-demo-ui-control-greeting"
      name="greeting"
    >
      <option
        value="Hello"
      >
        Hello
      </option>
      <option
        selected=""
        value="Bonjour"
      >
        Bonjour
      </option>
      <option
        value="„Åì„Çì„Å´„Å°„ÅØ"
      >
        „Åì„Çì„Å´„Å°„ÅØ
      </option>
    </select>
  </div>
  <div
    class="component-demo-ui-control-panel-label"
  >
    name
  </div>
  <div
    class="component-demo-ui-control-panel-control"
  >
    <input
      class="component-demo-ui-control"
      data-testid="component-demo-ui-control-name"
      name="name"
      value="Bob"
    />
  </div>
  <div
    class="component-demo-ui-control-panel-label"
  >
    hasEmoji
  </div>
  <div
    class="component-demo-ui-control-panel-control"
  >
    <label
      class="component-demo-ui-checkbox-control"
      data-testid="component-demo-ui-control-hasEmoji"
    >
      <input
        checked=""
        class="component-demo-ui-checkbox-control-input"
        name="hasEmoji"
        type="checkbox"
      />
    </label>
  </div>
</div>
`;

exports[`Preview should display the inspector 1`] = `
<div
  class="component-demo-ui-inspector"
  data-testid="component-demo-ui-inspector"
>
  <h3
    class="component-demo-ui-inspector-title"
  >
    Inspector
  </h3>
  <ol
    role="tree"
    style="padding: 0px; margin: 0px; list-style-type: none;"
  >
    <li
      aria-expanded="false"
      role="treeitem"
      style="color: black; background-color: white; line-height: 1.2; cursor: default; box-sizing: border-box; list-style: none; font-family: Menlo, monospace; font-size: 11px;"
    >
      <div>
        <span
          style="color: rgb(128, 128, 128);"
        >
          undefined
        </span>
      </div>
      <ol
        role="group"
        style="margin: 0px; padding-left: 12px;"
      />
    </li>
  </ol>
</div>
`;

exports[`Preview should provide an interface to interact with the inspector 1`] = `
<div
  class="component-demo-ui-inspector"
  data-testid="component-demo-ui-inspector"
>
  <h3
    class="component-demo-ui-inspector-title"
  >
    Inspector
  </h3>
  <ol
    role="tree"
    style="padding: 0px; margin: 0px; list-style-type: none;"
  >
    <li
      aria-expanded="false"
      role="treeitem"
      style="color: black; background-color: white; line-height: 1.2; cursor: default; box-sizing: border-box; list-style: none; font-family: Menlo, monospace; font-size: 11px;"
    >
      <div>
        <span
          style="color: rgb(110, 110, 110); display: inline-block; font-size: 12px; margin-right: 3px; user-select: none; transform: rotateZ(0deg);"
        >
          ‚ñ∂
        </span>
        <span
          style="font-style: italic;"
        />
        <span
          style="font-style: italic;"
        >
          {
          <span>
            <span
              style="color: rgb(136, 19, 145);"
            >
              hello
            </span>
            :¬†
            <span
              style="color: rgb(196, 26, 22);"
            >
              "
              World
              "
            </span>
          </span>
          }
        </span>
      </div>
      <ol
        role="group"
        style="margin: 0px; padding-left: 12px;"
      />
    </li>
  </ol>
</div>
`;
